integration of vyper compiler

evgeniuz opened the pull request:
Barebones version of vyper compilation, works only with `--all` flag for now.

Here's rough plan of how I intend to implement it: `compile.all`/`compile.necessary` will iterate over `COMPILERS` and for each extension find contracts with that extension and execute specified compiler for it. Both these functions are independent from Solidity, so they should work for Vyper as well. `compile.with_dependencies` will handle Solidity compilation as it is doing now, `compile.vyper` will have same interface, but will work with Vyper instead.

It's possible to separate `compile.vyper` into it's own module (i. e. vyper.js), but in that case I think would be good idea to separate `compile.with_dependencies` into it's own module as well and rename it to `compile.solidity`.

There is some processing of `source_path`s to be operating system independent that would be needed both for Solidity and Vyper, so I plan to extract them from `compile` function into utility functions to avoid code duplication.

List of things to do:

- [x]  get Vyper version and throw error if there are Vyper contracts, but Vyper is not available;
- [x]  add more logging (filenames that are being compiled, filenames where error occured (in addition to printing `stderr`);
- [x]  add multiple compiler support for `compile.necessary`;
- [x]  add tests.

Closes #1144

evgeniuz commented:
My main question right now is if this approach is ok and if it is should I refactor `compile` and `compile.with_dependencies` into solidity.js module inside `truffle-compile` package?

gnidan commented:
@evgeniuz One note for now:

Please put this stuff in a _new_ package, something like `truffle-vyper-compile`. The plan for `truffle-compile` is that it will eventually be renamed `truffle-solidity-compile`. This way we can keep the various languages separate, and eventually find commonalities to abstract similar behavior.

Thanks for pursuing this!

evgeniuz commented:
Refactored vyper support into `truffle-compile-vyper`. So current approach is following: I've added `forEachCompiler` helper function that essentially wraps whatever was in `compile.all` and `compile.necessary` and runs it for each defined extension/compiler. I've also added support for different extensions to `Profiler.updated` and `find_contracts` (this one should be completely backwards compatible).

gnidan commented:
Couple things:
- There should be no vyper-specific anything in `truffle-compile`. This determination logic should be moved to `truffle-workflow-compile`.
- Possibly make `truffle-compile` and `truffle-compile-vyper` separately define the file extension lookup, and just keep `truffle-workflow-compile` small and delegate.
- This all should be moved to target the `next` branch. I'll change the base branch here so you can see any merge conflicts more easily.

evgeniuz commented:
@gnidan there are few things that are in `truffle-compile` that I would like to reuse in vyper compilation: `compile.all` and `compile.necessary` are independent on solidity and will work for vyper (and any other compiler in future). `Profiler.updated` (which is part of `compile.necessary) also doesn't have anything solidity-specific, it just reads timestamp from JSON and compares it to source timestamp. If I move vyper processing to `truffle-workflow-compile` I won't be able to use that.

I can define `COMPILERS` in `truffle-workflow-compile` and pass that as an option to `compile.all/necessary`, will that be ok?

gnidan commented:
@evgeniuz the approach that we've discussed internally (@cgewecke and I) is that we should duplicate logic first, then identify commonality second. At that point we can abstract the Profiler and CompilerSupplier, etc. In the meantime, just feel free to copy those components but make them as Vyper-specific as you want.

Part of the reason for this is because `truffle-compile` is an integral component of Truffle that **cannot break**. So we have to tread lightly.

The generic interface for a compiler that I imagine is:
- `compile()`
- `comple.all()`
- `compile.necessary()`

See `truffle-external-compile` on the `next` break for another (much simpler) example.

gnidan commented:
(I will likely have more comments, so I apologize in advance if I forget something)

evgeniuz commented:
My main concern was to reduce duplication of code, but if that is ok for now that answers most of questions. I will update branch with that and rebase in on top of `next`.

evgeniuz commented:
Rebased on top of next branch, `truffle-compile` is now untouched, `truffle-workflow-compile`/`truffle-config` have minimal changes and all vyper support is in separate package. I've used `Profiler.update` by loading it from `truffle-compile` and copy-pasted `compile.all/compile.necessary/compile.display` from `truffle-compile` (marked this with TODO comment, as a candidate for refactoring later).

Now compilation starts if there are `.vy` files in contracts directory, compiles necessary files by default and all files if `--all` flag is passed.

evgeniuz commented:
@gnidan, can you please review integrations and suggest if there's anything to improve (or maybe someone else should do this?).

I also have a question about tests: should I write tests for this functionality and if yes, how to handle CI, as there should be vyper executable available for it to work.

gnidan commented:
> @gnidan, can you please review integrations and suggest if there's anything to improve (or maybe someone else should do this?).

Yep! I have a review I started actually, but I was waiting for you to indicate that it was less WIP :) I'll get to finishing up the review either tonight or tomorrow sometime.

> I also have a question about tests: should I write tests for this functionality and if yes, how to handle CI, as there should be vyper executable available for it to work.

Please write tests! Just set up mocha in the package. In the root directory, update `.travis.yml` to install Vyper (or preferably download the Vyper docker image). If you can't get it to work properly, let me know and I'll help investigate.

evgeniuz commented on packages/truffle-compile-vyper/index.js:
```diff
@@ -0,0 +1,175 @@
+// Execute vyper for single source file
+function execVyper(source_path, callback) {
+  // we need to execute compile with separate calls,
+  // as vyper does not guarantee output order
```
Created an issue for that (https://github.com/ethereum/vyper/issues/1021), this code can be simplified once that is fixed. Yes, the output is unpredictable and there's no way to find out which is which (well, technically JSON is easy to detect and runtime binary is shorter than binary, but that is not a way to rely on).

gnidan replied:
Ah, could you please make this change? You are probably the best person to do it, rather than leaving it as technical debt for later.

evgeniuz commented on packages/truffle-compile-vyper/index.js:
```diff
@@ -0,0 +1,175 @@
+
+// -------- End of common with truffle-compile --------
+
+// Check that vyper is available, save it's version
```
fixed.

evgeniuz commented on packages/truffle-compile-vyper/.travis.yml:
```diff
@@ -0,0 +1,25 @@
+        -e TRAVIS_PULL_REQUEST_BRANCH \
+        -e TRAVIS_BRANCH \
+        -e TEST \
+      truffle/ci:latest run_tests
```
removed.

evgeniuz commented on packages/truffle-compile-vyper/.gitignore:
```diff
@@ -0,0 +1,3 @@
+node_modules
+yarn.lock
+package-lock.json
```
removed.

evgeniuz commented on packages/truffle-compile-vyper/index.js:
```diff
@@ -0,0 +1,175 @@
+
+// Check that vyper is available, save it's version
+function checkVyper(callback) {
+  exec('vyper --version', function (err, stdout, stderr) {
```
I've considered this, but it looks like `ethereum/vyper` docker is not updated (it has very old version there and is not actually working). I've used snap version for development.

evgeniuz commented:
Fixed minor issues, added tests.

nemaniarjun commented:
What is the status of this one?! I plan on using Vyper for a college assignment and would love to use truffle with it? Can we get this merged ASAP? 

@evgeniuz If this is not merged soon, can you outline steps for how I can add it to my truffle manually or should I just use your fork for now?

I initially planned on using [truper](https://www.npmjs.com/package/truper) to get it working and it works for test, but not with migrations.

nemaniarjun commented on packages/truffle-compile-vyper/index.js:
```diff
@@ -0,0 +1,175 @@
+  version: null,
+};
+
+const VYPER_PATTERN = '**/*.vy';
```
Can we have `v.py` and `vyper.py` as patterns too? A lot of people keep them in `.py` for syntax highlighting.

evgeniuz replied:
Makes sense, will add support for these.

evgeniuz replied:
Added support for `.v.py` and `.vyper.py` files.

nemaniarjun commented on packages/truffle-compile-vyper/index.js:
```diff
@@ -0,0 +1,175 @@
+  });
+};
+
+// contracts_directory: String. Directory where .sol files can be found.
```
These comments might need changing :grin:

evgeniuz replied:
These comments are part of copy-pasted code from `truffle-compile`, I've left them as is (these will be probably refactored into some common place later).

evgeniuz commented:
> @evgeniuz If this is not merged soon, can you outline steps for how I can add it to my truffle manually or should I just use your fork for now?

You can use my fork, `develop` branch. It's based on top of `next` branch of truffle. Should be working out of the box, just place vyper contracts in `contracts` directory and have vyper executable in PATH (e. g. `snap install vyper --edge --devmode`)

evgeniuz commented:
@gnidan Can you please review changes/unit tests and comment if this is good for merge or anything else should be fixed?

evgeniuz commented:
Changed `execVyper` to only call `vyper` once for all formats. Formats/their parsing use one array, that way they can be added/reordered as necessary.
